cnxdb/migrations/20170911201035_transform_cnxml_to_html.py

```diff
@@ -0,0 +1,142 @@
+    FROM module_files
+    WHERE module_files.module_ident = b.module_ident
+      AND module_files.filename = b.filename
+      AND module_files.fileid != b.fileid
+      AND module_files.module_ident IN %s""",
+                   (module_idents, module_idents))
+
+    num_complete = 0
+    for batch in _batcher(to_transform, batch_size):
+        module_idents = tuple([i[1] for i in batch])
+        logger.debug('Transform module_idents {}'.format(module_idents))
+
+        for fileid, module_ident in batch:
+            cursor.execute("""\
+WITH index_cnxml AS (
+    SELECT files.fileid, file
+        FROM module_files NATURAL JOIN files
+        WHERE module_ident = %(module_ident)s
+          AND filename = 'index.cnxml'
+        LIMIT 1
+), transformed AS (
+    SELECT html_content(%(module_ident)s) AS content
+) INSERT INTO files
+    (file, media_type)
+    SELECT convert_to(transformed.content, 'utf-8'), 'text/xml'
+        FROM index_cnxml, transformed
+        WHERE char_length(substring(encode(file, 'escape')
+                                    FROM 'cnxml-version=.0.7.')) > 0
+          AND NOT EXISTS (
+            SELECT 1 FROM files
+                WHERE sha1 = sha1(transformed.content))
+RETURNING fileid""", {'module_ident': module_ident})
+            new_fileid = cursor.fetchall()
+            if new_fileid:
+                cursor.execute("""\
+UPDATE module_files SET fileid = %s
+WHERE fileid = %s""", (new_fileid[0][0], fileid))
+
+        cursor.connection.commit()
+        num_complete += len(batch)
+        percent_comp = num_complete * 100.0 / num_todo
+        elapsed = time.time() - start
+        remaining_est = elapsed * (num_todo - num_complete) / num_complete
+        est_complete = start + elapsed + remaining_est
+        logger.info('{:.1f}% complete '
+                    'est: "{}" ({})'.format(percent_comp,
+                                            time.ctime(est_complete),
+                                            timedelta(0, remaining_est)))
+
+    logger.info('Total runtime: {}'.format(timedelta(0, elapsed)))
+
+
+def down(cursor):
+    # Restore module_files save in datamigrations.index_cnxml_html
+    cursor.execute("""\
+WITH to_delete AS (
+    SELECT module_ident, fileid FROM module_files
+    WHERE NOT EXISTS (
+        SELECT 1 FROM datamigrations.index_cnxml_html b WHERE
+            b.fileid = module_files.fileid)
+      AND EXISTS (
+        SELECT 1 FROM datamigrations.index_cnxml_html b WHERE
+            b.module_ident = module_files.module_ident)
+      AND filename = 'index.cnxml.html'
+), update AS (
+UPDATE module_files
+    SET fileid = b.fileid
+    FROM datamigrations.index_cnxml_html b
+    WHERE module_files.module_ident = b.module_ident
+      AND module_files.fileid != b.fileid
+      AND module_files.filename = 'index.cnxml.html'
+)
+DELETE FROM files
+    WHERE EXISTS (
+        SELECT 1 FROM to_delete
+        WHERE to_delete.fileid = files.fileid);
```

Comment:
Can we add a `AND NOT EXISTS (SELECT 1 FROM module_files WHERE fileid = to_delete.fileid)`?

Reply:
hehe. Testing it now. That exact condition.

Reply:
(rollback has no progress reporting - and really can't, I'm just spoiled)